@xen-orchestra/backups/_cleanVm.js

```diff
@@ -358,9 +370,39 @@ exports.cleanVm = async function cleanVm(
     })
   }
 
-  const doMerge = () => {
-    const promise = asyncMap(toMerge, async chain => limitedMergeVhdChain(chain, { handler, onLog, remove, merge }))
-    return merge ? promise.then(sizes => ({ size: sum(sizes) })) : promise
+  const doMerge = async () => {
+    const metadataToBeUpdated = {}
+    await asyncMap(toMerge, async chain => {
+      const merged = await limitedMergeVhdChain(chain, { handler, onLog, remove, merge })
+      if (merged !== undefined) {
+        const metadataPath = vhdsToJSons[chain[0]] // all the chain should have the same metada file
+        metadataToBeUpdated[metadataPath] = true
+      }
+    })
+
+    // update the size in the metadata where there have been some merge
+    await asyncMap(Object.keys(metadataToBeUpdated), async metadataPath => {
+      const metadata = JSON.parse(await handler.readFile(metadataPath))
+
+      // update size stored in the metadata if possible
+      let size
+      try {
+        size = await computeVhdsSize(handler, metadata.vhds)
+        // only update if we are in a normal case of a vhd growing after merge
+      } catch (error) {
+        onLog(`failed to get size of ${metadataPath} after merge`, { error })
+      }
+      // write updated metadata file
+      // don't reduce the size : a vhd s should not be smaller after merge
+      if (metadata.size < size || metadata.size === undefined) {
+        metadata.size = size
+        try {
+          await handler.writeFile(metadataPath, JSON.stringify(metadata), { flags: 'w' })
+        } catch (error) {
+          onLog(`failed to update size in backup metadata ${metadataPath} after merge`, { error })
+        }
+      }
+    })
```

Comment:
Can this logic be deduplicated?

We have two pieces of code computing the size of VHDs and updating the metadata, is there any way we could group them?

A possible solution (out of the top of my head) could be to handle the merge first and then computing the size, only warning if a merge did not occurred.

What's your opinion on this?